### ISSUE TITLE
sp adds a location header when one is already present ?

### ISSUE THREAD

**Issue by landryb** [georchestra/georchestra#2193]
Dunno if it's linked to https://github.com/georchestra/georchestra/issues/1655#issuecomment-350277396 or #2176, but i managed to get into a state where geoserver would return a `location` header adding the `;jsessionid` arg - `Proxy.java` re-adds a location header on top in https://github.com/georchestra/georchestra/blob/master/security-proxy/src/main/java/org/georchestra/security/Proxy.java#L807 ?
```
adjustLocation called for request: /wxs/web/
adjustLocation found target: wxs for request: /wxs/web/
adjustLocation process header: http://localhost:9280/wxs/web/;jsessionid=B31A2AE0B33E28121705B4EC0802C91C?0
Test location header: http://localhost:9280/wxs/web/;jsessionid=B31A2AE0B33E28121705B4EC0802C91C?0 against: http://localhost:9280/wxs/
adjustLocation from: http://localhost:9280/wxs/web/;jsessionid=B31A2AE0B33E28121705B4EC0802C91C?0 to 
                  /wxs/web/;jsessionid=B31A2AE0B33E28121705B4EC0802C91C?0
```
and then the client receives two `Location:` headers, violating HTTP rfc:
```
< Location: http://localhost:9280/wxs/web/;jsessionid=B31A2AE0B33E28121705B4EC0802C91C?0                                                  
< Location: /wxs/web/;jsessionid=B31A2AE0B33E28121705B4EC0802C91C?0
```
To workaround this, i think Proxy.java should set the Location header instead of blindly adding one.

```
index 68b87e49d..a11ed0236 100644
--- a/security-proxy/src/main/java/org/georchestra/security/Proxy.java
+++ b/security-proxy/src/main/java/org/georchestra/security/Proxy.java
@@ -804,7 +804,7 @@ public class Proxy {
                 if (resolvedURI.toString().startsWith(baseURI.toString())) {
                     // proxiedResponse.removeHeader(locationHeader);
                     String newLocation = "/" + target + "/" + resolvedURI.toString().substring(baseURI.toString().length());
-                    finalResponse.addHeader("Location", newLocation);
+                    finalResponse.setHeader("Location", newLocation);
```

Playing with that and the geopublisher in geonetwork (which uses rest on geoserver), i managed to get into a state where geoserver wont recognize me anymore, even after logging out from cas:
```
2018-10-03 16:11:59,572 WARN [geoserver.security] - Failed login, user admin from 10.0.0.7, forwarded for 172.20.85.109
```
(the admin user here is a local-only, not-in-ldap user used only for rest POST bypassing the SP)

tried setting `useRelativeRedirects=false` on geoservers's tomcat but that didnt help.

**Comment by pmauduit** [comment]
I had a quick test on sdi-stable.g.o and even if I am pretty sure of the relevance of the modification (setHeader instead of addHeader), it still does not work on the demo instance. I Need to investigate further.

**Comment by groldan** [comment]
@landryb @pmauduit is there a concise set of steps to reproduce this? sorry if it's pretty obvious but I'm just getting to know the proxy implementation.

**Comment by landryb** [comment]
It's been a while, but i had to replace addHeader by setHeader on my production instance (i have a local commit since october), otherwise geoserver got into a broken state, as the SP returned two `Location` headers to the client (probably one from geoserver, and one added on top by the SP) which violates RFCs. Testing that it causes no regressions is what matters now :)

**Comment by groldan** [comment]
Problem is Proxy does `copyResponseHeaders` followed by `adjustLocation` and both add the `Location` header.

```
            headerManagement.copyResponseHeaders(request, request.getRequestURI(), proxiedResponse, finalResponse, this.targets);

            if (statusCode == 302 || statusCode == 301) {
                adjustLocation(request, proxiedResponse, finalResponse);
            }
```

 Changing to `finalResponse.setHeader()` as mentioned above makes no difference because the header was actually already sent if the implementation does not cache them. 

See the javadocs for `setHeader`:
```
     * Sets a response header with the given name and value.
     * If the header had already been set, the new value overwrites the
     * previous one.  The <code>containsHeader</code> method can be
     * used to test for the presence of a header before setting its
     * value.
```
I'm about to issue a PR.

**Comment by fvanderbiest** [comment]
On master I could reproduce using the docker composition with eg:
```
curl -Iksu testadmin:testadmin https://georchestra.mydomain.org/geoserver/web/wicket/bookmarkable/org.geoserver.web.data.layer.LayerPage | grep location
```

**Comment by groldan** [comment]
Fixed by #2413
